odxtools/response.py

```diff
@@ -23,32 +33,106 @@ class ResponseType(Enum):
+            for et_parameter in et_element.iterfind("PARAMS/PARAM")
+        ])
+        sdgs = [
+            SpecialDataGroup.from_et(sdge, doc_frags) for sdge in et_element.iterfind("SDGS/SDG")
+        ]
+
+        return Response(
+            response_type=response_type,
+            admin_data=admin_data,
+            parameters=parameters,
+            sdgs=sdgs,
+            **kwargs)
+
+    def _build_odxlinks(self) -> Dict[OdxLinkId, Any]:
+        result = {self.odx_id: self}
+
+        for param in self.parameters:
+            result.update(param._build_odxlinks())
+
+        return result
+
+    def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
+        for param in self.parameters:
+            param._resolve_odxlinks(odxlinks)
+
+    def _resolve_snrefs(self, context: SnRefContext) -> None:
+        context.parameters = self.parameters
+
+        try:
+            for param in self.parameters:
+                param._resolve_snrefs(context)
+        finally:
+            context.parameters = None
 
-    def encode(self, coded_request: Optional[bytes] = None, **kwargs: ParameterValue) -> bytes:
+    def encode(self, coded_request: Optional[bytes] = None, **kwargs: ParameterValue) -> bytearray:
         encode_state = EncodeState(triggering_request=coded_request, is_end_of_pdu=True)
 
         self.encode_into_pdu(physical_value=kwargs, encode_state=encode_state)
 
         return encode_state.coded_message
 
-    def _resolve_snrefs(self, context: SnRefContext) -> None:
-        context.response = self
+    def decode(self, message: bytes) -> ParameterValueDict:
+        decode_state = DecodeState(coded_message=message)
+        param_values = self.decode_from_pdu(decode_state)
+
+        if not isinstance(param_values, dict):
+            odxraise("Decoding a response must result in a dictionary")
+
+        return cast(ParameterValueDict, param_values)
+
+    def encode_into_pdu(self, physical_value: Optional[ParameterValue],
+                        encode_state: EncodeState) -> None:
+        composite_codec_encode_into_pdu(self, physical_value, encode_state)
+
+    def decode_from_pdu(self, decode_state: DecodeState) -> ParameterValue:
+        return composite_codec_decode_from_pdu(self, decode_state)
+
+    def get_static_bit_length(self) -> Optional[int]:
+        return composite_codec_get_static_bit_length(self)
+
+    @property
+    def required_parameters(self) -> List[Parameter]:
+        return composite_codec_get_required_parameters(self)
+
+    @property
+    def free_parameters(self) -> List[Parameter]:
+        return composite_codec_get_free_parameters(self)
+
+    def print_free_parameters_info(self) -> None:
+        """Return a human readable description of the structure's
+        free parameters.
+        """
+        from .parameterinfo import parameter_info
 
-        super()._resolve_snrefs(context)
+        print(parameter_info(self.free_parameters), end="")
```

Comment:
print?

Reply:
yes, IMO this is okay because the name of the calling method is `print_free_parameters_info()`. (it is meant to get overview of the parameters expected by `encode()`/produced by `decode()` during development. IOW it is not meant to be called during "normal" program execution.)